### ISSUE TITLE
Reboots not occurring on machine with etcd not on 127.0.0.1

### ISSUE THREAD

**Issue by crawford** [coreos/bugs#839]
<a href="https://github.com/adamalex"><img src="https://avatars.githubusercontent.com/u/50389?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [adamalex](https://github.com/adamalex)**
_Thursday Jul 03, 2014 at 12:41 GMT_
_Originally opened as https://github.com/coreos/locksmith/issues/23_

----

Hello!  You may remember me from https://github.com/coreos/fleet/issues/607 where I was getting assistance with securing a cluster to run on an isolated Rackspace network.  @jonboulle provided some expert guidance and I have things configured and running great on initial boot.  I am now finding that machines will not automatically reboot after updating.

After startup but before update downloaded, I found many of the following logged:

`Jul 03 11:16:52 corecluster locksmithd[3611]: Unlocking old locks failed: error setting up lock. Retrying in 5m0s.`

Then, the update appears to have downloaded successfully and the following loop is now occurring:

```
Jul 03 11:43:06 corecluster locksmithd[3611]: LastCheckedTime=1404387652 Progress=0 CurrentOperation="UPDATE_STATUS_UPDATED_NEED_REBOOT" NewVersion=0.0.0.0 NewSize=101901659
Jul 03 11:43:06 corecluster locksmithd[3611]: etcd.service is active
Jul 03 11:43:06 corecluster locksmithd[3611]: Error initializing etcd client: 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Jul 03 11:43:06 corecluster systemd[1]: locksmithd.service: main process exited, code=exited, status=1/FAILURE
Jul 03 11:43:06 corecluster systemd[1]: Unit locksmithd.service entered failed state.
Jul 03 11:43:16 corecluster systemd[1]: locksmithd.service holdoff time over, scheduling restart.
Jul 03 11:43:16 corecluster systemd[1]: Stopping Cluster reboot manager...
Jul 03 11:43:16 corecluster systemd[1]: Starting Cluster reboot manager...
Jul 03 11:43:16 corecluster systemd[1]: Started Cluster reboot manager.
Jul 03 11:43:16 corecluster locksmithd[4118]: locksmithd starting currentOperation="UPDATE_STATUS_UPDATED_NEED_REBOOT" strategy="best-effort"
Jul 03 11:43:16 corecluster locksmithd[4118]: etcd.service is active
Jul 03 11:43:16 corecluster locksmithd[4118]: Error initializing etcd client: 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Jul 03 11:43:16 corecluster systemd[1]: locksmithd.service: main process exited, code=exited, status=1/FAILURE
Jul 03 11:43:16 corecluster systemd[1]: Unit locksmithd.service entered failed state.
Jul 03 11:43:26 corecluster systemd[1]: locksmithd.service holdoff time over, scheduling restart.
Jul 03 11:43:26 corecluster systemd[1]: Stopping Cluster reboot manager...
Jul 03 11:43:26 corecluster systemd[1]: Starting Cluster reboot manager...
Jul 03 11:43:26 corecluster systemd[1]: Started Cluster reboot manager.
Jul 03 11:43:26 corecluster locksmithd[4125]: locksmithd starting currentOperation="UPDATE_STATUS_UPDATED_NEED_REBOOT" strategy="best-effort"
Jul 03 11:43:26 corecluster locksmithd[4125]: etcd.service is active
Jul 03 11:43:26 corecluster locksmithd[4125]: Error initializing etcd client: 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
```

My hypothesis is that I need to do one or the other of the following:

1. Configure locksmithd to communicate with etcd on the isolated network IP instead of the default localhost

2. Configure etcd to listen on 127.0.0.1 in addition to the isolated network IP, but not on 0.0.0.0 (to maintain the isolation I'm going after)

I wasn't able to find how to do either of the above, so I would appreciate any pointers you have on how to make that configuration or any other ideas you have to get past this.

Once this is all worked out I hope to provide a complete set of instructions for securing a CoreOS cluster by running within an isolated Rackspace network.

**Comment by crawford** [comment]
<a href="https://github.com/jonboulle"><img src="https://avatars.githubusercontent.com/u/2601015?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jonboulle](https://github.com/jonboulle)**
_Thursday Jul 03, 2014 at 19:19 GMT_

----

> Configure etcd to listen on 127.0.0.1 in addition to the isolated network IP, but not on 0.0.0.0 (to maintain the isolation I'm going after)

@unihorn is this possible? (today or in the future..)

**Comment by crawford** [comment]
<a href="https://github.com/philk"><img src="https://avatars.githubusercontent.com/u/29562?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [philk](https://github.com/philk)**
_Monday Jul 28, 2014 at 18:59 GMT_

----

I'm also having issues with this same thing. I've made etcd listen on the private network via some custom environment hackery but locksmithd and fleet are both unaware of that which causes them not to work.

**Comment by crawford** [comment]
<a href="https://github.com/jonboulle"><img src="https://avatars.githubusercontent.com/u/2601015?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jonboulle](https://github.com/jonboulle)**
_Monday Aug 04, 2014 at 22:34 GMT_

----

@philips any strong feelings on how to proceed here? Making locksmithd's etcd server configurable is easy enough but I dunno if I'm quite convinced it's the right path.

**Comment by crawford** [comment]
<a href="https://github.com/adamalex"><img src="https://avatars.githubusercontent.com/u/50389?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [adamalex](https://github.com/adamalex)**
_Monday Aug 04, 2014 at 23:28 GMT_

----

For some extra context, you can see the method I worked out for configuring etcd and fleet here:

https://gist.github.com/adamalex/eeed0825488b4831f559

I wasn't able to find any similar approach that would let me make the equivalent config to locksmith.

**Comment by crawford** [comment]
<a href="https://github.com/yichengq"><img src="https://avatars.githubusercontent.com/u/2976232?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [yichengq](https://github.com/yichengq)**
_Thursday Aug 07, 2014 at 08:37 GMT_

----

@jonboulle 
Yes, we could add the feature to let etcd listen on multiple addresses in the future. It is totally doable. This still needs discussion but is on TODO list from 0.5 to 0.6 I think.

**Comment by crawford** [comment]
<a href="https://github.com/kelseyhightower"><img src="https://avatars.githubusercontent.com/u/1123322?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kelseyhightower](https://github.com/kelseyhightower)**
_Monday Oct 06, 2014 at 21:17 GMT_

----

Has a decision been made on this issue? I think it would be nice to have a flag to set the etcd endpoint similar to what we have for fleet.

**Comment by crawford** [comment]
<a href="https://github.com/jonboulle"><img src="https://avatars.githubusercontent.com/u/2601015?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jonboulle](https://github.com/jonboulle)**
_Monday Oct 06, 2014 at 21:53 GMT_

----

Previously I wasn't convinced but after thinking about it more and considering the future direction of etcd I think it makes sense to add such a flag. Unless @philips feels strongly otherwise I'll do it.

**Comment by crawford** [comment]
<a href="https://github.com/adamalex"><img src="https://avatars.githubusercontent.com/u/50389?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [adamalex](https://github.com/adamalex)**
_Wednesday Oct 29, 2014 at 15:38 GMT_

----

I see it got merged in #38 -- thanks much!
